File: backend/app/Models/Producto.php
<?php
namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Producto extends Model
{
    protected $table = 'productos';

    protected $fillable = [
        'codigo',
        'codigo_barras',
        'nombre',
        'descripcion_ticket',
        'marca_id',
        'sub_marca_id',
        'categoria_id',
        'sub_categoria_id',
        'unidad_medida_id',
        'unidad_compra_id',
        'unidad_base_id',
        'factor_compra_base',
        'descripcion',
        'imagen',
        'ficha_tecnica',
        'accion_tecnica',
        'precio_base',
        'stock_minimo',
        'stock_maximo',
        'activo',
    ];

    /**
     * Siguiente código correlativo (PROD001, PROD002…) para cuando no se envía
     * uno. Toma el mayor número usado con ese prefijo y avanza hasta encontrar
     * uno libre, por si hay huecos o códigos escritos a mano.
     */
    public static function generarCodigo(string $prefijo = 'PROD'): string
    {
        $ultimo = static::where('codigo', 'like', $prefijo.'%')
            ->orderByRaw('LENGTH(codigo) DESC, codigo DESC')
            ->value('codigo');

        $n = (int) preg_replace('/\D/', '', (string) $ultimo);

        do {
            $n++;
            $codigo = $prefijo.str_pad((string) $n, 3, '0', STR_PAD_LEFT);
        } while (static::where('codigo', $codigo)->exists());

        return $codigo;
    }

    protected function casts(): array
    {
        return [
            'precio_base' => 'decimal:2',
            'factor_compra_base' => 'decimal:2',
            'stock_minimo' => 'decimal:2',
            'stock_maximo' => 'decimal:2',
            'activo' => 'boolean',
        ];
    }

    public function marca() { return $this->belongsTo(Marca::class); }
    public function subMarca() { return $this->belongsTo(SubMarca::class); }
    public function categoria() { return $this->belongsTo(Categoria::class); }
    public function subCategoria() { return $this->belongsTo(Categoria::class, 'sub_categoria_id'); }
    public function unidadMedida() { return $this->belongsTo(UnidadMedida::class); }
    public function unidadCompra() { return $this->belongsTo(UnidadMedida::class, 'unidad_compra_id'); }
    public function unidadBase() { return $this->belongsTo(UnidadMedida::class, 'unidad_base_id'); }

    public function presentaciones() { return $this->hasMany(ProductoPresentacion::class); }
    public function lotes() { return $this->hasMany(ProductoLote::class); }
    public function stocks() { return $this->hasMany(ProductoAlmacenStock::class); }
    public function movimientos() { return $this->hasMany(MovimientoInventario::class); }
}

File: app_abarrotes/lib/models/producto.dart
class Producto {
  final int id;
  final String codigo;
  final String nombre;
  final int marcaId;
  final int? subMarcaId;
  final int? categoriaId;
  final int unidadMedidaId;
  final String? descripcion;
  final double precioBase;
  final bool afectoIgv;
  final bool activo;
  final String? marcaNombre;
  final String? categoriaNombre;
  final String? unidadMedidaNombre;

  Producto({
    required this.id,
    required this.codigo,
    required this.nombre,
    required this.marcaId,
    this.subMarcaId,
    this.categoriaId,
    required this.unidadMedidaId,
    this.descripcion,
    this.precioBase = 0,
    this.afectoIgv = true,
    this.activo = true,
    this.marcaNombre,
    this.categoriaNombre,
    this.unidadMedidaNombre,
  });

  factory Producto.fromJson(Map<String, dynamic> json) {
    return Producto(
      id: json['id'],
      codigo: json['codigo'],
      nombre: json['nombre'],
      marcaId: json['marca_id'],
      subMarcaId: json['sub_marca_id'],
      categoriaId: json['categoria_id'],
      unidadMedidaId: json['unidad_medida_id'],
      descripcion: json['descripcion'],
      precioBase: (json['precio_base'] ?? 0).toDouble(),
      afectoIgv: json['afecto_igv'] ?? true,
      activo: json['activo'] ?? true,
      marcaNombre: json['marca']?['nombre'],
      categoriaNombre: json['categoria']?['nombre'],
      unidadMedidaNombre: json['unidad_medida']?['nombre'],
    );
  }

  Map<String, dynamic> toJson() => {
    'codigo': codigo,
    'nombre': nombre,
    'marca_id': marcaId,
    'sub_marca_id': subMarcaId,
    'categoria_id': categoriaId,
    'unidad_medida_id': unidadMedidaId,
    'descripcion': descripcion,
    'precio_base': precioBase,
    'afecto_igv': afectoIgv,
    'activo': activo,
  };
}
